File: packages/sdk-node/src/resources/invoices.ts
import type { TierstackHttpClient } from "../client";
import type { ListParams, Page, RequestOptions } from "../types";
import type { PaymentAttemptSummary } from "./subscriptions";

export type InvoiceStatus = "DRAFT" | "OPEN" | "PAID" | "VOID" | "UNCOLLECTIBLE";

export interface InvoiceLineItem {
  id: string;
  type: string;
  description: string;
  quantity: number;
  /** Minor units. Negative for credits and discounts. */
  unitAmount: number;
  amount: number;
  currency: string;
  periodStart: string | null;
  periodEnd: string | null;
}

export interface Invoice {
  id: string;
  organizationId: string;
  customerId: string;
  subscriptionId: string | null;
  invoiceNumber: string;
  status: InvoiceStatus;
  currency: string;
  subtotal: number;
  discountAmount: number;
  creditAmount: number;
  taxAmount: number;
  total: number;
  amountPaid: number;
  amountDue: number;
  billingPeriodStart: string | null;
  billingPeriodEnd: string | null;
  finalizedAt: string | null;
  paidAt: string | null;
  voidedAt: string | null;
  dunningAttempts: number;
  nextRetryAt: string | null;
  createdAt: string;
  /** Present only on `retrieve` — the list endpoint omits both for row size. */
  lineItems?: InvoiceLineItem[];
  attempts?: PaymentAttempt[];
}

export interface PaymentAttempt {
  id: string;
  invoiceId: string;
  customerId: string;
  provider: string;
  amount: number;
  currency: string;
  status: string;
  attemptNumber: number;
  failureCode: string | null;
  failureReason: string | null;
  providerReference: string | null;
  createdAt: string;
  completedAt: string | null;
}

export interface ListInvoicesParams extends ListParams {
  customerId?: string;
  subscriptionId?: string;
  status?: InvoiceStatus;
}

export interface PayInvoiceParams {
  /** Charge a specific stored method rather than the customer's default. */
  paymentMethodId?: string;
  /** Where the customer returns after a hosted checkout, if one is opened. */
  callbackUrl?: string;
  metadata?: Record<string, unknown>;
}

export class InvoicesResource {
  constructor(private readonly http: TierstackHttpClient) {}

  list(params: ListInvoicesParams = {}): Promise<Page<Invoice>> {
    return this.http.request("GET", "/v1/invoices", { query: params });
  }

  /** Includes line items and every payment attempt made against it. */
  retrieve(invoiceId: string): Promise<Invoice> {
    return this.http.request("GET", `/v1/invoices/${encodeURIComponent(invoiceId)}`);
  }

  /** Collect, or retry collecting, an open invoice. Every call creates a new attempt — previous ones are never overwritten. */
  pay(
    invoiceId: string,
    params: PayInvoiceParams = {},
    options?: RequestOptions
  ): Promise<PaymentAttemptSummary> {
    return this.http.request("POST", `/v1/invoices/${encodeURIComponent(invoiceId)}/pay`, {
      body: params,
      options,
    });
  }

  void(invoiceId: string): Promise<Invoice> {
    return this.http.request("POST", `/v1/invoices/${encodeURIComponent(invoiceId)}/void`);
  }
}

File: packages/sdk-python/src/tierstack/resources/invoices.py
from __future__ import annotations

from typing import Any, Literal, cast

from ..client import TierstackHttpClient
from ..types import TypedDict, compact
from .subscriptions import PaymentAttemptSummary

InvoiceStatus = Literal["DRAFT", "OPEN", "PAID", "VOID", "UNCOLLECTIBLE"]


class InvoiceLineItem(TypedDict):
    id: str
    type: str
    description: str
    quantity: int
    unitAmount: int
    """Minor units. Negative for credits and discounts."""
    amount: int
    currency: str
    periodStart: str | None
    periodEnd: str | None


class PaymentAttempt(TypedDict):
    id: str
    invoiceId: str
    customerId: str
    provider: str
    amount: int
    currency: str
    status: str
    attemptNumber: int
    failureCode: str | None
    failureReason: str | None
    providerReference: str | None
    createdAt: str
    completedAt: str | None


class Invoice(TypedDict):
    id: str
    organizationId: str
    customerId: str
    subscriptionId: str | None
    invoiceNumber: str
    status: InvoiceStatus
    currency: str
    subtotal: int
    discountAmount: int
    creditAmount: int
    taxAmount: int
    total: int
    amountPaid: int
    amountDue: int
    billingPeriodStart: str | None
    billingPeriodEnd: str | None
    finalizedAt: str | None
    paidAt: str | None
    voidedAt: str | None
    dunningAttempts: int
    nextRetryAt: str | None
    createdAt: str
    lineItems: list[InvoiceLineItem]
    """Present only on ``retrieve`` — the list endpoint omits both for row size."""
    attempts: list[PaymentAttempt]


class InvoicePage(TypedDict):
    items: list[Invoice]
    page: int
    limit: int
    total: int
    totalPages: int


class InvoicesResource:
    def __init__(self, http: TierstackHttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        customer_id: str | None = None,
        subscription_id: str | None = None,
        status: InvoiceStatus | None = None,
        q: str | None = None,
        page: int | None = None,
        limit: int | None = None,
    ) -> InvoicePage:
        query = {
            "customerId": customer_id,
            "subscriptionId": subscription_id,
            "status": status,
            "q": q,
            "page": page,
            "limit": limit,
        }
        return cast(InvoicePage, self._http.request("GET", "/v1/invoices", query=query))

    def retrieve(self, invoice_id: str) -> Invoice:
        """Includes line items and every payment attempt made against it."""
        return cast(Invoice, self._http.request("GET", f"/v1/invoices/{invoice_id}"))

    def pay(
        self,
        invoice_id: str,
        *,
        payment_method_id: str | None = None,
        callback_url: str | None = None,
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> PaymentAttemptSummary:
        """Collect, or retry collecting, an open invoice. Every call creates a
        new attempt — previous ones are never overwritten."""
        body = compact(
            {"paymentMethodId": payment_method_id, "callbackUrl": callback_url, "metadata": metadata}
        )
        return cast(
            PaymentAttemptSummary,
            self._http.request(
                "POST", f"/v1/invoices/{invoice_id}/pay", body=body, idempotency_key=idempotency_key
            ),
        )

    def void(self, invoice_id: str) -> Invoice:
        return cast(Invoice, self._http.request("POST", f"/v1/invoices/{invoice_id}/void"))
